## Laborationsövningar

Skriv ett uttryck som skriver ut ordet "kakaopulver" sjutton gånger i följd. Resultatet borde se ut så här: `kakaopulverkakaopulver[...]kakaopulver`.

In [26]:
# Din kod här
print("kakaopulver"*17)

kakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulverkakaopulver


Hur många tecken innehåller strängen?

In [ ]:
# Din kod här
len("kakaopulver"*17)

In [ ]:
# HINT (dubbelklicka här)
# Det går alldeles utmärkt att googla efter t ex "length of string in python"

En sträng, `"kakaopulver"` består av enskilda ordnade element som sammanfogats. Det innebär att vi kan indexera strängar på samma sätt som listor, med hakklamrar []. Skriv ut "kakaopulver" baklänges genom att använda använda en indexeringsmetod.

In [ ]:
# Skriv ut kakaopulver baklänges genom att lägga till rätt indexering i hakklamrarna nedan
"kakaopulver"[]

---
Dela upp följande mening i en lista med enskilda ord genom att använda en sträng-metod: `"The quick brown fox jumps over the lazy dog"`

In [30]:
# Din kod här
"The quick brown fox jumps over the lazy dog".split()

['The', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog']

In [31]:
# Dela nu upp den här strängen i en lista
"The_quick_brown_fox_jumps_over_the_lazy_dog"

# Din kod här
"The_quick_brown_fox_jumps_over_the_lazy_dog".split("_")

['The', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog']

---
Skriv en for-loop som skriver ut namnen på städerna i listan nedan om de innehåller ett "ö".

In [6]:
cities = ["stockholm", "malmö", "göteborg", "linköping", "norrköping", "västerås"]

In [7]:
# Din kod här
for city in cities:
    if "ö" in city:
        print(city)

malmö
göteborg
linköping
norrköping


Skriv nu en for-loop som skriver ut hur många bokstäver det är i varje stadsnamn samt namnet på staden. Första bokstaven i stadens namn ska vara en versal och namnet ska omges av citattecken. Resultatet bör se ut så här: 

I "Stockholm" finns det 9 bokstäver

I "Malmö" finns det 5 bokstäver

I "Göteborg" finns det 8 bokstäver

I "Linköping" finns det 9 bokstäver

I "Norrköping" finns det 10 bokstäver

I "Västerås" finns det 8 bokstäver

In [8]:
# Din kod här:
for city in cities:
    print(f"I \"{city.capitalize()}\" finns det {len(city)} bokstäver")

I "Stockholm" finns det 9 bokstäver
I "Malmö" finns det 5 bokstäver
I "Göteborg" finns det 8 bokstäver
I "Linköping" finns det 9 bokstäver
I "Norrköping" finns det 10 bokstäver
I "Västerås" finns det 8 bokstäver


In [32]:
# HINT
# Här krävs det att du använder en formaterad sträng.

---
Här nedanför ser du en oändlig while-loop. **Kör den inte än**. Lägg istället till en konstruktion som gör att den skriver ut "Klar!" och avbryts när k är 10.

In [ ]:
k = 0

while True:
    k += 1

In [36]:
# Lösningsförslag
k = 0

while True:
    k += 1
    if k == 10:
        print("Klar!")
        break

Klar!


---
Skriv en funktion `odd_or_even` med en parameter som skriver ut om det angivna värdet på parametern är jämnt eller udda.

In [ ]:
# Din kod här
def odd_or_even(n):
    if n % 2 == 0:
        print("Talet är jämnt!")
    else:
        print("Talet är udda!")

In [ ]:
# HINT
# Modulus-operatorn är väldigt användbar här. a % b visar vilken rest (om någon) som man får vid division av a och b.as_integer_ratio

In [ ]:
# Testa din funktion här
odd_or_even(1) # Bör skriva ut att värdet är udda
odd_or_even(6) # Bör skriva ut att värdet är jämnt

---
Använd ternära uttryck för att kolla om två objekt är lika. Skriv ut "It's a match" om `a` och `b` är lika, annars "No match".

In [2]:
a = 724379872437987243798724379872437987243798724379872437987243798724379872437987243798724379872437982743798
b = 724379872437987243798724379872437987243798724379872437987243798724379872437987243798724379872437987243798

# Din kod här
"It's a match" if a == b else "No match"

'No match'

---
Skriv en lambdafunktion med namn `printif` med en parameter `x`. Funktionen ska skriva ut *"Vilket stort tal!"* om värdet är **större** än 10, annars ska strängen *"`x` är ett lågt tal"* (där `x` är parameterns värde) skrivas ut. Du kan förutsätta att input till funktionen är numerisk.

In [ ]:
# Din funktion här
printif = lambda x: print(f"{x} är ett lågt tal") if x <= 10 else print("Vilket stort tal!")


In [37]:
# HINT
# Ett ternärt uttryck kan komma till nytta...

In [ ]:
# Testa din kod här.
# Första resultatraden borde vara: 10 är ett lågt tal
# Andra raden borde vara: Vilket stort tal!

printif(10)
printif(20)

Testa att köra cellen nedan. Vad blir det för fel? Varför?

In [ ]:
printif("två")

Definiera nu en funktion som heter `printer` utan några parametrar. Funktionen ska:
1. Låta användaren skriva in en valfri `input`.
1. Använda användarens input som argument till `printif`-funktionen
1. Innehålla minst ett `try-except`-block som försöker hantera uppkomna fel

Obs! `input()` returnerar en sträng, alltså behöver input-värdet konverteras. Om konverteringen misslyckas bör användaren informeras om att endast numeriska värden får anges.

In [ ]:
# Din funktion här:
def printer():
    x = input()
    try:
        printif(int(x))
    except ValueError:
        print("Conversion of input to int failed, only use numeric inputs please.")

In [ ]:
# Testa din funktion genom att ange några olika värden: 1, "one", [1], 0b10000.
printer()

In [ ]:
# HINT 1
# Typkonverteringen görs med int()

In [ ]:
# HINT 2
# Du kan göra konverteringen i ett try-block

---
Nu ska vi kasta tärning! 

1. Skapa en funktion `tärning` som representerar ett kast med en 6-sidig tärning, dvs. den ska returnera ett slumptal mellan 1 och 6. 
2. Skapa ytterligare en funktion `räkna_kast` som ska kalla på funktionen `tärning` och returnera antalet tärningskast som krävs för att få en etta. Tips: använd en while-loop i `räkna_kast`. 

Kalla på funktionen `tärning` från funktionen `räkna_kast`. 

**HINT:**
För att generera slumptal kan du importera modulen `random` genom att skriva följande i en cell: `import random`

Slumptal kan sen genereras med funktionen `random.randint([lägre_gräns], [övre_gräns])`.

In [3]:
# Din kod här
import random

def tärning():
    return random.randint(1, 6)


def räkna_kast(func):
    antal_kast = 1
    while func() != 1:
        antal_kast += 1
    return antal_kast

In [5]:
# Testa din funktion här
print(räkna_kast(tärning))

3


---
RPA-teamet utvecklar smarta robotar. Den smarta roboten Roger ska identifiera vilka system som behöver underhåll eller har kraschat. Listan `system` symboliserar en rad olika system, där varje objekt antar värdet "Grön", "Gul" eller "Röd" som representerar statusen på ett system. Skriv en funktion `hitta_fel` som tar en `färgkod` och en `lista` som parametrar, som ska printa alla index i listan för `färgkod`. Till exempel för färgkoden "Gul" så ska funktionen `hitta_fel` printa följande: Robot Roger har identifierat kod 'Gul' på position [0, 2, 4, 15]

Använd en for-loop och flödeskontroll. Om du vill får du gärna testa att använda en list comprehension istället för en for-loop. 

In [ ]:
# Kör den här cellen
system = ["Gul", "Grön", "Gul", "Grön", "Gul", "Röd", "Grön", "Röd", "Grön",
          "Grön", "Grön", "Grön", "Grön", "Röd", "Röd", "Gul", "Grön", "Grön", "Grön"]

In [6]:
# Din kod här
def hitta_index(färgkod, lista):
    index = list() 
    lista_längd = len(lista)
    # List comprehension
    # index = [i for i in range(lista_längd) if lista[i] == x]
    for i in range(lista_längd):
        if lista[i] == färgkod:
            index.append(i)

    print(f"Robot Roger har identifierat kod '{färgkod}' på position {index}")


In [7]:
# Testa din funktion här

# Se vilka system som behöver underhåll
hitta_index(färgkod="Gul", lista=system)

# Se vilka system som har kraschat
hitta_index(färgkod="Röd", lista=system)


Robot Roger har identifierat kod 'Gul' på position [0, 2, 4, 15]
Robot Roger har identifierat kod 'Röd' på position [5, 7, 13, 14]


---
Christer kör ofta bil till kontoret. Under resan noterar han alla epa-traktorer han ser och lägger i en lista. Han lägger även till alla motorcyklar, giraffer och gula bilar i listan.

En morgon fick Christer följande lista:

In [11]:
christers_morgonlista = ["motorcykel", "gul bil", "epa", "gul bil", "gul bil", "gul bil", "giraff",
                         "gul bil", "epa", "epa", "epa", "motorcykel", "epa", "epa", "gul bil", "epa", "epa", "epa", "epa", "motorcykel", "epa", "epa"]


Hjälp Christer att skapa en dictionary `epa_dict` som innehåller antalet av varje objekt i `christers_morgonlista`.

In [11]:
# Din kod här
epa_dict = {}

for vehicle in range(len(christers_morgonlista)):
    epa_dict[christers_morgonlista[vehicle]] = christers_morgonlista.count(christers_morgonlista[vehicle])

print(epa_dict)

{'motorcykel': 3, 'gul bil': 6, 'epa': 12, 'giraff': 1}


En korrekt skriven dictionary ska innehålla följande: {'motorcykel': 3, 'gul bil': 6, 'giraff': 1, 'epa': 12}. Obs! Ordningen kan skilja.

In [22]:
# HINT
christers_morgonlista.count("motorcykel")

3

In [16]:
# Lösningsförslag 2: använd set för att få de unika objekten i listan
epa_dict = {}

for item in set(christers_morgonlista):
    epa_dict[item] = christers_morgonlista.count(item)

print(epa_dict)

{'gul bil': 6, 'motorcykel': 3, 'giraff': 1, 'epa': 12}


In [21]:
# Lösningsförslag 3: dictionary comprehension
epa_dict = {item: christers_morgonlista.count(item) for item in set(christers_morgonlista)}
epa_dict

{'gul bil': 6, 'motorcykel': 3, 'giraff': 1, 'epa': 12}

Definera en funktion `printa_dict` som tar en dictionary som parameter. Funktionen ska skriva ut samtliga nycklar och värden genom att loopa igenom `epa_dict`. Använd formaterade strängar. Nyckeln "epa" ska skrivas ut med versaler. Resultatet bör se ut enligt följande:


motorcykel: 3

gul bil: 6

giraff: 1

EPA: 12

In [13]:
# Din kod här

def printa_dict(d):
    for objekt, antal in d.items():
        if "epa" in objekt:
            print(f"{objekt}: {antal}".upper())
        else:
            print(f"{objekt}: {antal}")


In [14]:
# Testa din funktion här
printa_dict(epa_dict)

motorcykel: 3
gul bil: 6
EPA: 12
giraff: 1


(Överkurs) Sortera nycklarna i `epa_dict` efter dess värden (i fallande ordning) och spara som `sortera_epa_dict`. Detta genom att kombinera en dictionary comprehension med en lambdafunktion. 

In [15]:
# Din kod här:
sortera_epa_dict = {key: antal for (key, antal) in sorted(epa_dict.items(), key=lambda x: x[1], reverse=True)}
print(sortera_epa_dict.items())

dict_items([('epa', 12), ('gul bil', 6), ('motorcykel', 3), ('giraff', 1)])


In [ ]:
# HINT 
# Du kan använda dig av funktionen sorted()

In [16]:
# Titta på din sorterade dictionary här
printa_dict(sortera_epa_dict)

EPA: 12
gul bil: 6
motorcykel: 3
giraff: 1


---
Vi älskar burgare! Definera en klass med klassnamnet `Burgare`. Klassen ska ha:

1. En konstruktor
2. Attributen gram, topping och vegetarisk. Gram är en integer (default `90`), topping är en lista med godsaker (default `["ost"]`, vegetarisk kan anta de booleska värdena `True` eller `False`)
3. Metod `pris` som tillhör klassen Burgare som ska räkna ut priset för en burgare.


Metoden pris ska göra följande:
* Ta en rabattkupong som parameter
* Burgare som är mindre än 120 gram kostar 120 kronor. Övriga kostar 150 kronor.
* Varje topping kostar 15 kronor
* Vegetariskt alternativ kostar 20 kronor mindre än icke-vegetariskt alternativ.

In [23]:
# Din kod här
class Burgare:
    """Modellerar en burgare"""

    # Konstructor
    def __init__(self, gram = 90, topping = ["ost"], vegetarisk = False):
        """Konstruktor"""
        self.gram = gram
        self.topping = topping
        self.vegetarisk = vegetarisk

    # Methods
    def pris(self, rabattkupong):
        """Beräkna pris för burgare"""
        total = 0
        if self.gram < 120:
            total += 120
        else:
            total += 150

        total += len(self.topping) * 15

        if self.vegetarisk:
            total -= 20

        total *= 1 - rabattkupong 
        return total

In [24]:
# Testa din funktion här:
ingen_rabatt = 0
rabatt25 = 0.25

en_burgare = Burgare(gram = 150, topping = ["ost", "sallad", "lök", "dressing"], vegetarisk = False)
en_veg_burgare = Burgare(gram = 90, topping = ["ost", "tomat", "dressing"], vegetarisk = True)

print(f"Pris för en burgare med {ingen_rabatt}% rabatt: {en_burgare.pris(ingen_rabatt)} kronor") 
print(f"Pris för en burgare med {int(rabatt25*100)}% rabatt: {en_burgare.pris(rabatt25)} kronor") 

print(f"Pris för en vegetarisk burgare med {ingen_rabatt}% rabatt: {en_veg_burgare.pris(ingen_rabatt)} kronor") 
print(f"Pris för en vegetarisk burgare med {int(rabatt25*100)}% rabatt: {en_veg_burgare.pris(rabatt25)} kronor") 

Pris för en burgare med 0% rabatt: 210 kronor
Pris för en burgare med 25% rabatt: 157.5 kronor
Pris för en vegetarisk burgare med 0% rabatt: 145 kronor
Pris för en vegetarisk burgare med 25% rabatt: 108.75 kronor


**Koden ovan ska returnera följande:**

- Pris för en burgare med 0% rabatt: 210 kronor
- Pris för en burgare med 25% rabatt: 157.5 kronor
- Pris för en vegetarisk burgare med 0% rabatt: 145 kronor
- Pris för en vegetarisk burgare med 25% rabatt: 108.75 kronor